File: src/services/timeline.ts
import type { Incident, TimelineEvent } from "@/types/incident";

/** Timeline from facts the source actually supplied. No invented events. */
export function buildTimeline(i: Incident): TimelineEvent[] {
  const ev: TimelineEvent[] = [{ at: i.dispatchedAt, kind: "created", text: `Incident created — ${i.callType}` }];
  for (const u of i.units) {
    if (u.dispatchedAt) {
      const later = u.dispatchedAt.getTime() - i.dispatchedAt.getTime() > 60_000;
      ev.push({ at: u.dispatchedAt, kind: later ? "unit_added" : "unit_dispatched", text: `${u.unit} ${later ? "added" : "dispatched"}`, unit: u.unit });
    }
  }
  if (i.updatedAt && i.updatedAt.getTime() > i.dispatchedAt.getTime()) ev.push({ at: i.updatedAt, kind: "updated", text: "Incident updated" });
  if (i.closedAt) ev.push({ at: i.closedAt, kind: "closed", text: "Incident closed" });
  return ev.sort((a, b) => a.at.getTime() - b.at.getTime());
}

File: api/mcfrs_api/services/timeline.py
"""Timeline from facts the source actually supplied, merged with what this
instance observed. No invented events."""

from __future__ import annotations

from datetime import timedelta

from mcfrs_api.models import Incident, TimelineEvent


def source_timeline(i: Incident) -> list[TimelineEvent]:
    ev = [TimelineEvent(at=i.dispatched_at, kind="created", text=f"Incident created — {i.call_type}")]
    for u in i.units:
        if u.dispatched_at:
            later = u.dispatched_at - i.dispatched_at > timedelta(minutes=1)
            ev.append(
                TimelineEvent(
                    at=u.dispatched_at,
                    kind="unit_added" if later else "unit_dispatched",
                    text=f"{u.unit} {'added' if later else 'dispatched'}",
                    unit=u.unit,
                )
            )
    if i.updated_at and i.updated_at > i.dispatched_at:
        ev.append(TimelineEvent(at=i.updated_at, kind="updated", text="Incident updated"))
    if i.closed_at:
        ev.append(TimelineEvent(at=i.closed_at, kind="closed", text="Incident closed"))
    return sorted(ev, key=lambda e: e.at)


def merge_timeline(source: list[TimelineEvent], observed: list[TimelineEvent]) -> list[TimelineEvent]:
    """Observed events add detail (unit added at HH:MM as seen by us); duplicates of a
    source-supplied event within a minute are dropped."""
    out = list(source)
    for o in observed:
        dup = any(s.kind == o.kind and s.unit == o.unit and abs((s.at - o.at).total_seconds()) < 60 for s in source)
        if not dup:
            out.append(o)
    return sorted(out, key=lambda e: e.at)
